### PR TITLE
Added/Edited new chapter/assembly about foreman webhooks

### PR DESCRIPTION
**Pull request by ianf77**
Edited previous work on foreman webhooks.

[doc] Foreman Webhooks

https://issues.redhat.com/browse/SAT-3106

Cherry-pick into:

* [x] Foreman 2.5 (Satellite 6.10)
* [ ] Foreman 2.4
* [ ] Foreman 2.3 (Satellite 6.9)
* [ ] Foreman 2.1 (Satellite 6.8)

### CODE REVIEW

**Comment by ianf77** [comment]
Sorry, that hasn't worked as I wanted. It's not showing he assembly/procedures etc.

**Comment by ianf77** [comment]
All sorted now I hope.

File: guides/doc-Administering_Red_Hat_Satellite/build/en-US/Author_Group.xml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (melcorr) [inline guides/doc-Administering_Red_Hat_Satellite/build/en-US/Author_Group.xml:1]: This file  shouldn't be here.

File: guides/doc-Administering_Red_Hat_Satellite/build/en-US/Revision_History.xml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (melcorr) [inline guides/doc-Administering_Red_Hat_Satellite/build/en-US/Revision_History.xml:1]: This file shouldn't be here.

File: guides/doc-Administering_Red_Hat_Satellite/build/en-US/master.ent [hunk 3]
```diff
@@ -0,0 +1,4 @@
+<!ENTITY PRODUCT "Red Hat Satellite">
```
Comment (melcorr) [inline guides/doc-Administering_Red_Hat_Satellite/build/en-US/master.ent:1]: This file shouldn't be here.

**Review by melcorr — CHANGES_REQUESTED** [review]
Hi ian, 
Thanks for working on this.
I haven't completed a review but there is a lot of xml files that should not be here at all. 
What are the pngs and the SVG for? 
Looks like the procedures need intro context to them also.

**Comment by ianf77** [comment]
These files came as a surprise I admit. I don't know why they're here. 
If I try to make BUILD the Admin guide it complains about failing to create symlinks for all the png files. It's the only guide that does that, and always has.

**Comment by melcorr** [comment]
make clean before you build again
Please clean all this up.
You might need to configure a .gitignore

**Comment by ianf77** [comment]
Cleaned up unrelated files as per comments.

File: guides/common/modules/proc_creating-a-webhook-template.adoc [hunk 4]
```diff
@@ -0,0 +1,12 @@
+[id="creating-a-webhook-template_{context}"]
+= Creating a Webhook Template
+
+.Procedure
+
+. In the {Project} web UI, navigate to *Administer* > *Webhooks Templates*.
```
Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook-template.adoc:6]:
```suggestion
. In the ProjectWebUI, navigate to *Administer* > *Webhooks Templates*.
```
We have a proper attribute for this: :ProjectWebUI: {https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-base.adoc#L84
Please apply throughout.

File: guides/common/modules/proc_creating-a-webhook.adoc [hunk 5]
```diff
@@ -0,0 +1,43 @@
+[id="creating-a-webhook_{context}"]
+= Creating a Webhook
+
```
Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook.adoc:3]: Modules would always have some context.

File: guides/common/modules/proc_creating-a-webhook.adoc [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+. Click *Subscribe to field* to reveal a list of available events.
+. Select an event from the list.
+. Enter a name.
+. Enter a target URL. When using the shellhooks plugin, the URL should be in the form of `https://{smartproxy.example.com}:{smartproxy_port}/shellhooks/my_script`.
+. Click *Template* to see the list of available templates.
+. Select a template from the list.
+. Enter a HTTP method.
+. Ensure that the *Enabled* flag is checked if you want to create active webhook.
+
+In the Credentials tab:
```
Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook.adoc:17]: We do not document by tab. 
Please rewrite with the step "Click the Credentials tab" and then integrate the rest of the steps as steps on their own. 
We write user-oriented not web ui oriented.

File: guides/common/modules/proc_creating-a-webhook.adoc [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+. Optional: Enter the username and password when HTTP authentication is required. Do not enter anything when using shellhooks.
+. Select *Verify SSL* if the server certificate should be verified against the system certificate store or Foreman CA. When using shellhooks, ensure this flag is selected.
+. Select *Proxy Authorization* when using shellhooks, otherwise deselect this box.
+. Optional: Provide one or more CA certificates in the X509 Certification Authorities. Keep this field blank when using shellhooks.
+
+In the Additional tab:
```
Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook.adoc:24]: Same here

File: guides/common/modules/ref_webhooks-templates.adoc [hunk 8]
```diff
@@ -0,0 +1,12 @@
+[id="webhooks-templates_{context}"]
+= Templates
```
Comment (melcorr) [inline guides/common/modules/ref_webhooks-templates.adoc:2]: This title is extremely vague. 
Mod docs style would require a descriptive title. Descriptive titles help with search engine optimisation. 
I think this document would be really benefit from Ingrid's minimalism session. @ryandeussing do you think you could help @ianf77 set this up for Ingrid's group to review this document? It would be a hugely beneficial learning experience.

File: guides/common/modules/ref_webhooks-templates.adoc [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
```
Comment (melcorr) [inline guides/common/modules/ref_webhooks-templates.adoc:3]: The text in this section is without any developed context, and reads like a group of random statements. 
As a user, why am I here reading this? 
Some structure would be beneficial.

File: guides/common/modules/ref_webhooks-templates.adoc [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+The webhook HTTP payload must be created using {Project} template syntax.
+The subject of the event is available using the `@object` variable.
+{Project} ships with an example template for every single event available.
+For more information, see https://access.redhat.com/documentation/en-us/red_hat_satellite/6.9/html/managing_hosts/appe-red_hat_satellite-managing_hosts-template_writing_reference[Template Writing Reference] in _Managing Hosts_ and for available template macros and methods, visit `/templates_doc` on {ProjectServer}.
```
Comment (melcorr) [inline guides/common/modules/ref_webhooks-templates.adoc:7]: Please use the attributes for all links.

File: guides/common/modules/proc_installing-webhooks-shellhooks-plugin.adoc [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+[options="nowrap" subs="+quotes,attributes"]
+----
+# foreman-installer --enable-foreman-plugin-webhooks
+----
+
+* Optionally, install the shellhooks plugin on each {SmartProxy} that is used for shellhooks, using the following command:
++
+[options="nowrap" subs="+quotes,attributes"]
+----
+# foreman-installer --enable-foreman-proxy-plugin-shellhooks
```
Comment (melcorr) [inline guides/common/modules/proc_installing-webhooks-shellhooks-plugin.adoc:18]: You need to add an attribute for foreman-installer
https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-base.adoc#L2

File: guides/common/modules/ref_migrating-from-foreman_webhooks.adoc [hunk 12]
```diff
@@ -0,0 +1,12 @@
+[id="migrating-webhooks_{context}"]
+= Migrating from Foreman Webhooks
```
Comment (melcorr) [inline guides/common/modules/ref_migrating-from-foreman_webhooks.adoc:2]: This title makes it look like it is a procedure. 
Gerunds are used for procedures. 
However I cannot find the procedure in the text. 
Please develop a title to best suit the content and respect the mod doc and minimalist conventions.

File: guides/common/modules/ref_shellhooks.adoc [hunk 13]
```diff
@@ -0,0 +1,10 @@
+[id="shellhooks_{context}"]
+= Shellhooks
+
+Optionally, {SmartProxy} Shellhooks plugin can be installed which exposes executables using a REST HTTPS API.
+Scripts must be placed in `/var/lib/foreman-proxy/shellhooks` as executables with only alphanumeric characters and underscores in the name.
+The HTTPS payload is passed using standard input, optional command line arguments can be provided using X-Shellhook-Arg-1 _to_ N.
+HTTP method must be POST, an example URL would be: `https://{smartproxy.example.com}:{smartproxy_port}/shellhooks/my_script`.
```
Comment (melcorr) [inline guides/common/modules/ref_shellhooks.adoc:7]: This is an incorrect attribute. 
Please reference the attributes-base.adoc file and update.

File: guides/common/modules/proc_creating-a-webhook.adoc [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+. Enter the HTTP Content Type, typically application/json, application/xml or text/plain depending on the payload.
+. Optional: Provide HTTP headers as JSON. ERB is allowed and the same macros and variables are available as in webhook templates.
+
+When configuring webhooks with endpoints with non-standard HTTP(s) ports, SELinux port must be assigned, see  https://access.redhat.com/documentation/en-us/red_hat_satellite/6.9/html-single/installing_satellite_server_from_a_connected_network/index#configuring-selinux-to-ensure-access-on-custom-ports_satellite[Configuring SELinux to Ensure Access to Satellite on Custom Ports] in _Installing Satellite Server From A Connected Network_.
```
Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook.adoc:29]: We do not use (s) in IBM style: 

IBM style page 241:

Do not form a plural by adding (s). Try to rewrite the sentence to use either the plural
form or singular form, whichever is more appropriate. If you must indicate both forms,
repeat the noun, or use one or more. For example, instead of writing “enter the registra-
tion number(s),” write either “enter the registration number or numbers” or “enter one or
more registration numbers.”

Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook.adoc:29]: Please use the correct attributes for the links and test them locally.

Comment (melcorr) [reply guides/common/modules/proc_creating-a-webhook.adoc:29]: Please do not use Satellite Server. Use the attributes.

File: guides/common/assembly_using-foreman-webhooks.adoc [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+The body of a webhook request, also known as payload, can be customized through an associated Webhook template and can be any HTTP content type.
+HTTP authentication, content type, and headers can also be configured. {ProjectServer} does not support receiving webhooks.
+
+include::modules/proc_installing-webhooks-shellhooks-plugin.adoc[leveloffset=+1]
+
+include::modules/proc_creating-a-webhook-template.adoc[leveloffset=+1]
+
+include::modules/proc_creating-a-webhook.adoc[leveloffset=+1]
+
+include::modules/ref_webhooks-available-events.adoc[leveloffset=+1]
```
Comment (melcorr) [inline guides/common/assembly_using-foreman-webhooks.adoc:24]: This should be listed last.

File: guides/common/modules/proc_creating-a-webhook.adoc [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+. Enter a name.
+. Enter a target URL. When using the shellhooks plugin, the URL should be in the form of `https://{smartproxy.example.com}:{smartproxy_port}/shellhooks/my_script`.
+. Click *Template* to see the list of available templates.
+. Select a template from the list.
+. Enter a HTTP method.
+. Ensure that the *Enabled* flag is checked if you want to create active webhook.
+
+In the Credentials tab:
+
+. Optional: Enter the username and password when HTTP authentication is required. Do not enter anything when using shellhooks.
```
Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook.adoc:19]: Please use one sentence per line, as is our agreed style.

File: guides/common/modules/proc_creating-a-webhook.adoc [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+. Select *Verify SSL* if the server certificate should be verified against the system certificate store or Foreman CA. When using shellhooks, ensure this flag is selected.
```
Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook.adoc:20]: Please use one sentence per line. As per our agreed style. This will not effect the formatting.

File: guides/common/modules/proc_creating-a-webhook.adoc [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+. Select *Proxy Authorization* when using shellhooks, otherwise deselect this box.
+. Optional: Provide one or more CA certificates in the X509 Certification Authorities. Keep this field blank when using shellhooks.
```
Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook.adoc:22]: Please use one sentence per line.

File: guides/common/modules/proc_creating-a-webhook.adoc [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+
+In the Additional tab:
+
+. Enter the HTTP Content Type, typically application/json, application/xml or text/plain depending on the payload.
+. Optional: Provide HTTP headers as JSON. ERB is allowed and the same macros and variables are available as in webhook templates.
```
Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook.adoc:27]: Please use one sentence per line.

File: guides/common/modules/proc_creating-a-webhook.adoc (hunk 14 above)
Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook.adoc:29]: Please use one sentence per line.

File: guides/common/modules/proc_creating-a-webhook.adoc [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+
+[options="nowrap" subs="+quotes,attributes"]
+----
+{
+  "X-Shellhook-Arg-1": "<%= @object.content_view_version_id %>,
+  "X-Shellhook-Arg-2": "<%= @object.content_view_name %>
+}
+----
+
+Ensure the content renders to a valid JSON. Also, only pass safe fields like database ID, name or labels which do not include new lines or quote characters.
```
Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook.adoc:43]: PLease use one sentence per line.

File: guides/common/modules/proc_creating-a-webhook.adoc [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+= Creating a Webhook
+
+.Procedure
+
+. In the {Project} web UI, click *Administer* > *Webhooks*.
+. Click *Create Webhook*.
+. Click *Subscribe to field* to reveal a list of available events.
+. Select an event from the list.
+. Enter a name.
+. Enter a target URL. When using the shellhooks plugin, the URL should be in the form of `https://{smartproxy.example.com}:{smartproxy_port}/shellhooks/my_script`.
```
Comment (melcorr) [inline guides/common/modules/proc_creating-a-webhook.adoc:11]: Please use one sentence per line.

**Review by melcorr — CHANGES_REQUESTED** [review]
Hi @ianf77 
Thanks for putting together and updating Suyog's initial draft. 
From an upstream/downstream attribute and IBM style perspective, I have added a number of comments. 
The general layout of this chapter does not really correspond with the rest of the docs. 
These reference files with their titles and info without context for the user are not standard.
The modules do not have context set or brief short descriptions. From what I remember, this is not ideal, can you correct me if I am wrong @yetanothertw ?
There is a lot of passive voice in this text, which sometimes is unavoidable, but something that I know we made an active effort to avoid when I was on the team. Sometimes reading through the reference modules I'm unsure whether this is something the user needs to do or just some general information. 
I think bringing this doc to Ingrid Towey's minimalism workshop will help. Especially when it comes to those reference files and the titles. 
For a long time now, we have a one line per sentence rule. Sergei brought this to the CCS Satellite group and we added this to our workflow a while back. Please ensure that you follow this. 

TL;DR if you apply my feedback, it's enough for the upstream community, but if you want to follow CCS standards, minimalism standards, and proper mod doc standards, please use this as a learning opportunity with Ingrid Towey's group. It would help everything now and in the future. 

cc @ryandeussing

File: guides/common/modules/ref_shellhooks.adoc [hunk 22]
```diff
@@ -0,0 +1,10 @@
...
+Authorization with Foreman client checkbox must be enabled for each webhook to authorize a call.
```
Comment (melcorr) [inline guides/common/modules/ref_shellhooks.adoc:8]:
```suggestion
Authorization with {Project} client checkbox must be enabled for each webhook to authorize a call.
```

**Comment by yetanothertw** [comment]
Hi folks, 
Hi @melcorr - regarding the use of the context variable, it is usually set at the assembly level, so that every module with that assembly resolves with that value at build time. 
You are correct that in this particular instance, the context variable is not set at all and must be added. 
I've pointed @ianf77 to the [Modular Documentation Reference Guide](https://redhat-documentation.github.io/modular-docs/#) for more information and an example of the the [assembly template](https://raw.githubusercontent.com/redhat-documentation/modular-docs/master/modular-docs-manual/files/TEMPLATE_ASSEMBLY_a-collection-of-modules.adoc) 

With regard to the other type of context, which is introductory sentences in any procedure modules, those comments are valid as well and require addressing before they can be deemed compliant with mod docs standards.

@ianf77 and I went briefly over these comments to make sure they were clear in terms of corrective actions that are needed to address them and to make sure Ian is supported with all the necessary links to additional resources. @ianf77 , I'm happy to answer any other questions, should you require additional onboarding resources.

I love @melcorr 's Minimalism classes suggestion (attendance and also submitting content for review), and I think it's a reasonable approach to take. Ingrid is so friendly, open to suggestions, and incredibly helpful - her classes are a fantastic learning opportunity.

Hope this helps!
Many thanks,
Vlada

**Comment by melcorr** [comment]
Thank you @yetanothertw 
@ianf77 - I can provide any further context you need to complete this
For engineering related questions, @lzap is great. 
The videos from the demo and from the Foreman Birthday Party that we shared with you last week might also help you frame the information from a user perspective a bit better. 
i would also like to point you to our readme files:
* https://github.com/theforeman/foreman-documentation/tree/master/guides#readme
* https://github.com/theforeman/foreman-documentation/blob/master/guides/README.md that has a lot of useful information. Perhaps if you were not familiar with working in GitHub, you might not have realised that readmes in GitHub repos contain valuable information about the projects and their standards. 
If you see something that is out of date, please raise a PR to update it. 
I am here if you have further questions!

**Comment by lzap** [comment]
If there are technical questions in this PR, please @lzap me otherwise unsubscribing :-)

File: guides/common/assembly_using-foreman-webhooks.adoc [hunk 23]
```diff
@@ -0,0 +1,33 @@
+ifdef::context[:parent-context: {context}]
+
+[id="using-foreman-webhooks_{context}"]
+= Using Foreman Webhooks
+
+A webhook is a way for a web page or web application to provide other applications with information.
+The information is provided immediately. Specific events trigger customized callbacks, such as sending an email when a user registers with a website, or acknowledging online purchases.
```
Comment (maximiliankolb) [inline guides/common/assembly_using-foreman-webhooks.adoc:7]:
```suggestion
The information is provided immediately.
Specific events trigger customized callbacks, such as sending an email when a user registers with a website, or acknowledging online purchases.
```

File: guides/common/assembly_using-foreman-webhooks.adoc [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+Because webhooks use HTTP, no new infrastructure need be added to existing web services.
+
+When the `foreman_webhooks` plugin is installed, {ProjectServer} can be configured to send webhook requests.
+Webhooks are only triggered after an event occurs and are only fire-and-forget as responses are ignored.
+The typical use case for webhooks in Foreman is making a call to a monitoring system when a host is created or deleted.
```
Comment (maximiliankolb) [inline guides/common/assembly_using-foreman-webhooks.adoc:12]:
```suggestion
The typical use case for webhooks in {Project} is making a call to a monitoring system when a host is created or deleted.
```

File: guides/common/assembly_using-foreman-webhooks.adoc [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+{ProjectServer} and various plugins broadcast internal events which, if they are of interest, can be subscribed to.
+Webhook objects can be created where the target URL that expressed interest and other details of the event are specified.
+The body of a webhook request, also known as payload, can be customized through an associated Webhook template and can be any HTTP content type.
+HTTP authentication, content type, and headers can also be configured. {ProjectServer} does not support receiving webhooks.
```
Comment (maximiliankolb) [inline guides/common/assembly_using-foreman-webhooks.adoc:16]:
```suggestion
HTTP authentication, content type, and headers can also be configured.
{ProjectServer} does not support receiving webhooks.
```

File: guides/common/modules/proc_creating-a-webhook.adoc [hunk 26]
```diff
@@ -0,0 +1,43 @@
+[id="creating-a-webhook_{context}"]
+= Creating a Webhook
+
+.Procedure
+
+. In the {Project} web UI, click *Administer* > *Webhooks*.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_creating-a-webhook.adoc:6]:
```suggestion
. In the {ProjectWebUI}, navigate to *Administer* > *Webhooks*.
```

File: guides/common/modules/proc_creating-a-webhook.adoc (hunk 21 above)
Comment (maximiliankolb) [inline guides/common/modules/proc_creating-a-webhook.adoc:11]:
```suggestion
. Enter a target URL.
When using the shellhooks plugin, the URL should be in the form of `https://{smartproxy-example-com}:{smartproxy_port}/shellhooks/my_script`.
```

one sentence per line and the attribute uses dashes, not dots. See [base attributes L95](https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-base.adoc#L95)

File: guides/common/modules/proc_creating-a-webhook.adoc (hunk 14 above)
Comment (maximiliankolb) [inline guides/common/modules/proc_creating-a-webhook.adoc:29]:
```suggestion
When configuring webhooks with endpoints with non-standard HTTP(s) ports, SELinux port must be assigned, see  https://access.redhat.com/documentation/en-us/red_hat_satellite/{AccessRedHatComVersion}/html-single/installing_satellite_server_from_a_connected_network/index#configuring-selinux-to-ensure-access-on-custom-ports_satellite[Configuring SELinux to Ensure Access to Satellite on Custom Ports] in _Installing Satellite Server From A Connected Network_.
```

Additionally, use an asciidoc attribute to link to access.redhat.com. See [attributes L17](https://github.com/theforeman/foreman-documentation/blame/master/guides/common/attributes.adoc#L17)

**Review by maximiliankolb — CHANGES_REQUESTED** [review]
Thanks for the PR. Please update your PR and see my suggestions.

File: guides/common/modules/proc_creating-a-webhook.adoc (hunk 14 above)
Comment (melcorr) [reply guides/common/modules/proc_creating-a-webhook.adoc:29]: He should just be using attributes L17 rather than adding {AccessRedHatComVersion} in this case. That is for Red Hat URLs that are not upstream guides, if I understand correctly.

Comment (maximiliankolb) [reply guides/common/modules/proc_creating-a-webhook.adoc:29]: My bad. If the content is available upstream, then yes. It's probably something like `{InstallingProjectDocURL}`.

**Comment by ryandeussing** [comment]
This document is under CCS Peer Review to address the issues and topics raised above. Thank you!

**Comment by ianf77** [comment]
Hi @melcorr 
I've updated the files after a peer review and incorporated the comments that were made before.

File: guides/common/modules/ref_webhooks-available-events.adoc [hunk 27]
```diff
@@ -1,6 +1,8 @@
 [id="webhooks-available-events_{context}"]
 = Available Events
 
+The following are the  webhook events available from the {ProjectWebUI}.
```
Comment (melcorr) [inline guides/common/modules/ref_webhooks-available-events.adoc:4]: Double space.

**Review by melcorr — COMMENTED** [review]
Before I review, all attributes must be applied. 
I cannot see updates to the URLs so they can be reused upstream and downstream. 
I commented throughout my previous review as to where these updates need to be made.

**Comment by melcorr** [comment]
Please close this if you do not intend to work on it any more.
